ndb/src/google/cloud/ndb/query.py

```diff
@@ -129,43 +140,729 @@ def resolve(self, bindings, used):
+        self._name = name
+        self._opsymbol = opsymbol
+        self._value = value
+        return self
+
+    def __getnewargs__(self):
+        """Private API used to specify ``__new__`` arguments when unpickling.
+
+        .. note::
+
+            This method only applies if the ``pickle`` protocol is 2 or
+            greater.
+
+        Returns:
+            Tuple[str, str, Any]: A tuple containing the
+            internal state: the name, ``opsymbol`` and value.
+        """
+        return self._name, self._opsymbol, self._value
+
+    def __repr__(self):
+        return "{}({!r}, {!r}, {!r})".format(
+            self.__class__.__name__, self._name, self._opsymbol, self._value
+        )
+
+    def __eq__(self, other):
+        if not isinstance(other, FilterNode):
+            return NotImplemented
+
+        return (
+            self._name == other._name
+            and self._opsymbol == other._opsymbol
+            and self._value == other._value
+        )
+
+    def _to_filter(self, post=False):
+        """Helper to convert to low-level filter, or :data:`None`.
+
+        Args:
+            post (bool): Indicates if this is a post-filter node.
+
+        Returns:
+            None: If this is a post-filter.
+
+        Raises:
+            NotImplementedError: If the ``opsymbol`` is ``!=`` or ``in``, since
+                they should correspond to a composite filter. This should
+                never occur since the constructor will create ``OR`` nodes for
+                ``!=`` and ``in``
+            NotImplementedError: If not a post-filter and the ``opsymbol``
+                is a simple comparison. (For now) this is because the original
+                implementation relied on a low-level datastore query module.
+        """
+        if post:
+            return None
+        if self._opsymbol in (_NE_OP, _IN_OP):
+            raise NotImplementedError(
+                "Inequality filters are not single filter "
+                "expressions and therefore cannot be converted "
+                "to a single filter ({!r})".format(self._opsymbol)
+            )
+
+        raise NotImplementedError("Missing datastore_query.make_filter")
 
 
 class PostFilterNode(Node):
-    def __init__(self, *args, **kwargs):
-        raise NotImplementedError
+    """Tree node representing an in-memory filtering operation.
+
+    This is used to represent filters that cannot be executed by the
+    datastore, for example a query for a structured value.
+
+    Args:
+        predicate (Callable[[Any], bool]): A filter predicate that
+            takes a datastore entity (typically as a protobuf) and
+            returns :data:`True` or :data:`False` if the entity matches
+            the given filter.
+    """
+
+    __slots__ = ("predicate",)
+
+    def __new__(cls, predicate):
+        self = super(PostFilterNode, cls).__new__(cls)
+        self.predicate = predicate
+        return self
```

Comment:
Is `self` conventional inside `__new__`?  I've seen `instance` elsewhere, I think.

Reply:
I have no relevant experience with this `__new__` constructor madness. I'm happy to rename to `instance` if you'd like.

Reply:
Sure, it will be a kindness to our future selves.

Reply:
Done. I did a cursory search for `self =`, LMK if you see any remaining?